Unify POST handling

nwc10 opened the pull request:
* refactor and unify the `POST` code from the Authenticator class with the `POST` and `GET` code in the Request code
* paranoid error handling - the different API endpoints use different formats to report errors - "don't cross the streams" is less of a problem now
* move `auth_host` to  Request, Authenticator only uses `host` internally
* use Test::Warnings in all tests

nwc10 committed:
Paranoid error handling with clear reporting in _authenticate

*We*, not user code, are handling errors from the external API here - it's
much easier to debug "the impossible" if we report it clearly, rather than
explode with Moose constraint violations. Given that we now have much more
detailed error reporting (both location and probable cause) it seems better
to use croak() rather than confess() - this lets the caller choose whether
they want to show/log a stack backtrace, rather than use forcing it on them.

Both success and error responses *should* be JSON, so we test for that
before deciding whether we are handling 2xx status codes or 4xx/5xx.

The original code was assuming the v1/v2 API error response format.

This still seems to be used in the API used to get an access token, but the
v3 API TrueLayer has changed to the JSON format described in RFC 7807.
(The migration guide doesn't mention this.) Handle both formats.

nwc10 committed:
Use Test::Warnings in all tests

No need to explicitly depend on (or require) Test::Exception as Test::Most
pulls it in implicitly.

nwc10 committed:
Move the post in _authenticate() to _ua_request() in the Request class

Refactor it so that 'POST' is a parameter, calling build_tx() and start().
This way it can also be used to merge the GET code (and any future DELETE
code).

Add _process_response_new() which will soon replace _process_response().

nwc10 committed:
Improve tests for Business::TrueLayer::Request prior to refactoring

nwc10 committed:
Refactor api_post() and api_get() to use _ua_request()

This eliminates _process_response() and inlines _process_response_new() into
_ua_request(). It changes the format of error messages slightly, so adapt
the tests. Remove the TODO - GET failures now report 'GET' not 'POST'.

api_post() and api_get() now take a parameter to explicitly state whether
the API is expected to return JSON, with the default being true. This
replaces the previous code which would heuristically treat an empty body
(or, '0') as "return nothing" and always return the value otherwise.

It's unclear whether we need this for either GET or POST - of all the
APIs I could see, only those using DELETE could return an empty body.
(With a 200 code on success, not a 204, which feels wrong.)

nwc10 committed:
`auth_host` moves to Request, and Authenticator only uses `host` internally

nwc10 commented:
I didn't update the `CHANGES` file - should I?
I didn't run the tests in `xt` - I wasn't quite sure what I needed to put into `TRUELAYER_CREDENTIALS` and it doesn't seem to be documented

leejo commented:
I'll bump the `Changes` file as well as running the `xt` tests 👍

leejo commented:
```

leejo commented:
```
/home/leejohnson/working/business-truelayer > TRUELAYER_CREDENTIALS=business-truelayer-credentials.json prove -lr xt/
xt/000_changes.t ................... ok
xt/001_authenticate.t .............. ok
xt/002_sign.t ...................... TrueLayer POST https://api.truelayer-sandbox.com/test-signature returned 204 with no MIME type at xt/002_sign.t line 21.
xt/002_sign.t ...................... Dubious, test returned 255 (wstat 65280, 0xff00)
No subtests run
xt/003_merchant_account_details.t .. ok
xt/004_create_payment.t ............ ok

Test Summary Report
-------------------
xt/002_sign.t                    (Wstat: 65280 (exited 255) Tests: 0 Failed: 0)
  Non-zero exit status: 255
  Parse errors: No plan found in TAP output
Files=5, Tests=17,  4 wallclock secs ( 0.02 usr  0.00 sys +  1.45 cusr  0.10 csys =  1.57 CPU)
Result: FAIL
```

Just need to tweak for `204` content types. Will do that then merge this

leejo committed:
handle 204 No Content responses

namely there being no content and not even a content_type header, tweak
test coverage to exercise this along with the ->test_signature method
